## 1) Week 1:Data exploration

In [0]:
import pandas as pd
import numpy as np
from google.colab import files
from google.colab import drive
import seaborn as sn
import matplotlib.pyplot as plt



In [3]:

drive.mount('/content/drive', force_remount=True)

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive


In [0]:
import os
os.chdir("/content/drive/My Drive")

In [0]:
transaction_train = pd.read_csv('train_transaction.csv')
identity_train = pd.read_csv('train_identity.csv')

[link text](https://)a) Description of dataset  'transaction_train'

In [6]:
transaction_train.describe(include= 'all')

TransactionID        isFraud  ...           V338           V339
count    5.905400e+05  590540.000000  ...   82351.000000   82351.000000
unique            NaN            NaN  ...            NaN            NaN
top               NaN            NaN  ...            NaN            NaN
freq              NaN            NaN  ...            NaN            NaN
mean     3.282270e+06       0.034990  ...     151.160542     100.700882
std      1.704744e+05       0.183755  ...    1095.034387     814.946722
min      2.987000e+06       0.000000  ...       0.000000       0.000000
25%      3.134635e+06       0.000000  ...       0.000000       0.000000
50%      3.282270e+06       0.000000  ...       0.000000       0.000000
75%      3.429904e+06       0.000000  ...       0.000000       0.000000
max      3.577539e+06       1.000000  ...  104060.000000  104060.000000

[11 rows x 394 columns]

Let us discover the number of column and row

In [7]:
print(transaction_train.shape)

(590540, 394)


We have 394 columns with 590541 rows.
Let us explore the columns names.

In [8]:
transaction_train.columns

Index(['TransactionID', 'isFraud', 'TransactionDT', 'TransactionAmt',
       'ProductCD', 'card1', 'card2', 'card3', 'card4', 'card5',
       ...
       'V330', 'V331', 'V332', 'V333', 'V334', 'V335', 'V336', 'V337', 'V338',
       'V339'],
      dtype='object', length=394)

In [9]:
transaction_train.head(n=10)

TransactionID  isFraud  TransactionDT  ...  V337 V338  V339
0        2987000        0          86400  ...   NaN  NaN   NaN
1        2987001        0          86401  ...   NaN  NaN   NaN
2        2987002        0          86469  ...   NaN  NaN   NaN
3        2987003        0          86499  ...   NaN  NaN   NaN
4        2987004        0          86506  ...   0.0  0.0   0.0
5        2987005        0          86510  ...   NaN  NaN   NaN
6        2987006        0          86522  ...   NaN  NaN   NaN
7        2987007        0          86529  ...   NaN  NaN   NaN
8        2987008        0          86535  ...   0.0  0.0   0.0
9        2987009        0          86536  ...   NaN  NaN   NaN

[10 rows x 394 columns]

NaN value 

In [12]:
transaction_train.isnull().sum()

TransactionID          0
isFraud                0
TransactionDT          0
TransactionAmt         0
ProductCD              0
                   ...  
V335              508189
V336              508189
V337              508189
V338              508189
V339              508189
Length: 394, dtype: int64

We have 590541 rows, but v335,v336, v337, v338 and v339 have 508189 Nan. We decide to remove these columns.

In [16]:
transaction_train['card1'].isnull()

0         False
1         False
2         False
3         False
4         False
          ...  
590535    False
590536    False
590537    False
590538    False
590539    False
Name: card1, Length: 590540, dtype: bool

In [27]:
transaction_train.drop(['V335', 'V336', 'V337', 'V338', 'V339'], axis =1)

TransactionID  isFraud  TransactionDT  ...  V332 V333  V334
0             2987000        0          86400  ...   NaN  NaN   NaN
1             2987001        0          86401  ...   NaN  NaN   NaN
2             2987002        0          86469  ...   NaN  NaN   NaN
3             2987003        0          86499  ...   NaN  NaN   NaN
4             2987004        0          86506  ...   0.0  0.0   0.0
...               ...      ...            ...  ...   ...  ...   ...
590535        3577535        0       15811047  ...   NaN  NaN   NaN
590536        3577536        0       15811049  ...   NaN  NaN   NaN
590537        3577537        0       15811079  ...   NaN  NaN   NaN
590538        3577538        0       15811088  ...   NaN  NaN   NaN
590539        3577539        0       15811131  ...   NaN  NaN   NaN

[590540 rows x 389 columns]

Let us verify again the Nan in dataframe

In [42]:
for i in transaction_train.columns:
  print(i, transaction_train[i].isnull().sum().sum())

TransactionID 0
isFraud 0
TransactionDT 0
TransactionAmt 0
ProductCD 0
card1 0
card2 8933
card3 1565
card4 1577
card5 4259
card6 1571
addr1 65706
addr2 65706
dist1 352271
dist2 552913
P_emaildomain 94456
R_emaildomain 453249
C1 0
C2 0
C3 0
C4 0
C5 0
C6 0
C7 0
C8 0
C9 0
C10 0
C11 0
C12 0
C13 0
C14 0
D1 1269
D2 280797
D3 262878
D4 168922
D5 309841
D6 517353
D7 551623
D8 515614
D9 515614
D10 76022
D11 279287
D12 525823
D13 528588
D14 528353
D15 89113
M1 271100
M2 271100
M3 271100
M4 281444
M5 350482
M6 169360
M7 346265
M8 346252
M9 346252
V1 279287
V2 279287
V3 279287
V4 279287
V5 279287
V6 279287
V7 279287
V8 279287
V9 279287
V10 279287
V11 279287
V12 76073
V13 76073
V14 76073
V15 76073
V16 76073
V17 76073
V18 76073
V19 76073
V20 76073
V21 76073
V22 76073
V23 76073
V24 76073
V25 76073
V26 76073
V27 76073
V28 76073
V29 76073
V30 76073
V31 76073
V32 76073
V33 76073
V34 76073
V35 168969
V36 168969
V37 168969
V38 168969
V39 168969
V40 168969
V41 168969
V42 168969
V43 168969
V44 168969
V45 16

We can remove column which has the great number of Nan

In [0]:
transaction_new =transaction_train.drop(['D6', 'D7', 'D8', 'D9','V322', 'V323', 'V324', 'V325', 'V326', 'V327', 'V328', 'V329',
                        'V330', 'V331', 'V332', 'V333', 'V334', 'V335',
                        'V336', 'V337', 'V338', 'V339'], axis =1)

In [47]:
transaction_new.head(5)

TransactionID  isFraud  TransactionDT  ...  V319 V320  V321
0        2987000        0          86400  ...   0.0  0.0   0.0
1        2987001        0          86401  ...   0.0  0.0   0.0
2        2987002        0          86469  ...   0.0  0.0   0.0
3        2987003        0          86499  ...   0.0  0.0   0.0
4        2987004        0          86506  ...   0.0  0.0   0.0

[5 rows x 376 columns]


## 2) Week 2: We continue to  explore the dataset.
 we want to remove completely column which has at least 508589 Nan. There are around 20 columns.  

In [0]:
for i in transaction_new.columns:
  number_Nan = transaction_new[i].isnull().sum().sum()
  if number_Nan >= 508589:
    transaction_new = transaction_new.drop([i], axis = 1)

In [54]:
transaction_new.head(5)

TransactionID  isFraud  TransactionDT  ...  V319 V320  V321
0        2987000        0          86400  ...   0.0  0.0   0.0
1        2987001        0          86401  ...   0.0  0.0   0.0
2        2987002        0          86469  ...   0.0  0.0   0.0
3        2987003        0          86499  ...   0.0  0.0   0.0
4        2987004        0          86506  ...   0.0  0.0   0.0

[5 rows x 339 columns]

Let us verify again the number of Nan

In [55]:
for i in transaction_new.columns:
  print(i, transaction_new[i].isnull().sum().sum())

TransactionID 0
isFraud 0
TransactionDT 0
TransactionAmt 0
ProductCD 0
card1 0
card2 8933
card3 1565
card4 1577
card5 4259
card6 1571
addr1 65706
addr2 65706
dist1 352271
P_emaildomain 94456
R_emaildomain 453249
C1 0
C2 0
C3 0
C4 0
C5 0
C6 0
C7 0
C8 0
C9 0
C10 0
C11 0
C12 0
C13 0
C14 0
D1 1269
D2 280797
D3 262878
D4 168922
D5 309841
D10 76022
D11 279287
D15 89113
M1 271100
M2 271100
M3 271100
M4 281444
M5 350482
M6 169360
M7 346265
M8 346252
M9 346252
V1 279287
V2 279287
V3 279287
V4 279287
V5 279287
V6 279287
V7 279287
V8 279287
V9 279287
V10 279287
V11 279287
V12 76073
V13 76073
V14 76073
V15 76073
V16 76073
V17 76073
V18 76073
V19 76073
V20 76073
V21 76073
V22 76073
V23 76073
V24 76073
V25 76073
V26 76073
V27 76073
V28 76073
V29 76073
V30 76073
V31 76073
V32 76073
V33 76073
V34 76073
V35 168969
V36 168969
V37 168969
V38 168969
V39 168969
V40 168969
V41 168969
V42 168969
V43 168969
V44 168969
V45 168969
V46 168969
V47 168969
V48 168969
V49 168969
V50 168969
V51 168969
V52 168969
V53 

Now the maximum number of Nan in a column is around 450909. What can we do with these columns? Do we remove each row corresponding to Nan value? What can we really do in this case.
We decide to remove column with proportion of missing value is around 0.70.


In [91]:
for i in transaction_new.columns:
  percent = transaction_new[i].isnull().sum().sum()/transaction_new.shape[0]
  
  if percent > 0.70 :
    transaction_new = transaction_new.drop([i], axis = 1)
transaction_new.head(7)

TransactionID  isFraud  TransactionDT  ...  V319 V320  V321
0        2987000        0          86400  ...   0.0  0.0   0.0
1        2987001        0          86401  ...   0.0  0.0   0.0
2        2987002        0          86469  ...   0.0  0.0   0.0
3        2987003        0          86499  ...   0.0  0.0   0.0
4        2987004        0          86506  ...   0.0  0.0   0.0
5        2987005        0          86510  ...   0.0  0.0   0.0
6        2987006        0          86522  ...   0.0  0.0   0.0

[7 rows x 220 columns]

Now, we want to remove column which has nan (Question: Is it correct?)

In [93]:
transaction_new = transaction_new.dropna()
transaction_new.shape

(76416, 220)

In [94]:
transaction_new.head(7)

TransactionID  isFraud  TransactionDT  ...   V319   V320   V321
9         2987009        0          86536  ...    0.0    0.0    0.0
36        2987036        0          87149  ...    0.0    0.0    0.0
41        2987041        0          87220  ...  774.0  774.0  774.0
46        2987046        0          87285  ...    0.0    0.0    0.0
51        2987051        0          87328  ...    0.0    0.0    0.0
73        2987073        0          87759  ...    0.0    0.0    0.0
86        2987086        0          87950  ...    0.0    0.0    0.0

[7 rows x 220 columns]

Week 3: This week, we search for relationship between the variables. We will make some plot and analyze them.